File: im-ai-office/backend-ts/src/repos.ts
import { and, desc, eq, ne, isNotNull, sql } from "drizzle-orm";
import { db } from "./db/drizzle.js";
import { aiDm, alias, audit, message, person, task, term } from "./db/schema.js";

// 数据访问层（Drizzle 查询构建器；行为与手写 SQL 版逐字等价，返回键保持 snake_case）

// ============ 人 / 别名 ============

export interface Person { id: number; real_name: string | null; flower_name: string | null; }

export async function findPersonsByAlias(name: string): Promise<Person[]> {
  return db.select({
    id: person.id,
    real_name: person.realName,
    flower_name: person.flowerName,
  })
    .from(alias)
    .innerJoin(person, eq(person.id, alias.personId))
    .where(eq(alias.name, name));
}

export async function distinctAliasNames(): Promise<string[]> {
  const rows = await db.selectDistinct({ name: alias.name }).from(alias);
  return rows.map((r) => r.name!);
}

export async function aliasLabelRows(): Promise<Array<{ name: string; real_name: string | null; flower_name: string | null }>> {
  return db.selectDistinct({
    name: alias.name,
    real_name: person.realName,
    flower_name: person.flowerName,
  })
    .from(alias)
    .innerJoin(person, eq(person.id, alias.personId)) as Promise<Array<{ name: string; real_name: string | null; flower_name: string | null }>>;
}

export async function insertAliasIfAbsent(personId: number, name: string): Promise<boolean> {
  const inserted = await db.insert(alias)
    .values({ personId, name })
    .onConflictDoNothing()
    .returning({ id: alias.id });
  return inserted.length > 0;
}

// ============ 任务 ============

export interface TaskRow {
  id: number; content: string; creator: string | null; assignee: string | null;
  deadline: string | null; deadline_at: Date | string | null; status: string | null;
  confidence: string | null; source_msg: string | null; pending_meta: string | null;
  created_at: Date | string; updated_at: Date | string | null;
}

type TaskSelect = {
  id: typeof task.id; content: typeof task.content; creator: typeof task.creator;
  assignee: typeof task.assignee; deadline: typeof task.deadline; deadline_at: typeof task.deadlineAt;
  status: typeof task.status; confidence: typeof task.confidence; source_msg: typeof task.sourceMsg;
  pending_meta: typeof task.pendingMeta; created_at: typeof task.createdAt; updated_at: typeof task.updatedAt;
};
const TASK_COLS: TaskSelect = {
  id: task.id, content: task.content, creator: task.creator,
  assignee: task.assignee, deadline: task.deadline, deadline_at: task.deadlineAt,
  status: task.status, confidence: task.confidence, source_msg: task.sourceMsg,
  pending_meta: task.pendingMeta, created_at: task.createdAt, updated_at: task.updatedAt,
};

export async function insertTask(
  content: string, creator: string, assignee: string | null, deadline: string | null,
  status: string, confidence: string, sourceMsg: string, pendingMeta: string | null = null
): Promise<number> {
  const rows = await db.insert(task)
    .values(pendingMeta !== null
      ? { content, creator, assignee, deadline, status, confidence, sourceMsg, pendingMeta }
      : { content, creator, assignee, deadline, status, confidence, sourceMsg })
    .returning({ id: task.id });
  return rows[0].id;
}

export async function getTaskDict(taskId: number): Promise<TaskRow | null> {
  const rows = await db.select(TASK_COLS).from(task).where(eq(task.id, taskId)).limit(1);
  return (rows[0] as unknown as TaskRow) ?? null;
}

export async function listTaskDicts(status?: string): Promise<TaskRow[]> {
  if (status) {
    return db.select(TASK_COLS).from(task).where(eq(task.status, status)).orderBy(desc(task.id)) as unknown as Promise<TaskRow[]>;
  }
  return db.select(TASK_COLS).from(task).where(ne(task.status, "cancelled")).orderBy(desc(task.id)) as unknown as Promise<TaskRow[]>;
}

export async function latestPendingAssigneeForCreator(creator: string): Promise<TaskRow | null> {
  const rows = await db.select(TASK_COLS).from(task)
    .where(and(eq(task.creator, creator), eq(task.status, "pending_assignee")))
    .orderBy(desc(task.id)).limit(1);
  return (rows[0] as unknown as TaskRow) ?? null;
}

export async function latestPendingAssigneeByDmTaskid(senderId: string): Promise<TaskRow | null> {
  const dm = await db.select({ task_id: aiDm.taskId }).from(aiDm)
    .where(and(eq(aiDm.senderId, senderId), isNotNull(aiDm.taskId)))
    .orderBy(desc(aiDm.id)).limit(1);
  if (!dm.length || dm[0].task_id == null) return null;
  const rows = await db.select(TASK_COLS).from(task)
    .where(and(eq(task.id, dm[0].task_id), eq(task.status, "pending_assignee")))
    .orderBy(desc(task.id)).limit(1);
  return (rows[0] as unknown as TaskRow) ?? null;
}

// ============ 消息表 ============

export interface MessageRow {
  id: number; conv_id: string; sender_id: string; sender_name: string; content: string;
  content_type: number; is_self: number; msg_seq: number | null; client_msg_id: string | null; ts: Date | string;
}

/** 幂等：同 conv 内相同 client_msg_id 已存在时直接返回既有 id。 */
export async function messageAdd(
  convId: string, senderId: string, senderName: string, content: string,
  isSelf = 0, msgSeq: number | null = null, clientMsgId: string | null = null, contentType = 101
): Promise<number> {
  if (clientMsgId) {
    const hit = await db.select({ id: message.id }).from(message)
      .where(and(eq(message.convId, convId), eq(message.clientMsgId, clientMsgId)))
      .limit(1);
    if (hit.length) return Number(hit[0].id);
  }
  const rows = await db.insert(message)
    .values({ convId, senderId, senderName, content, isSelf, msgSeq, clientMsgId, contentType })
    .returning({ id: message.id });
  return rows[0].id;
}

export async function messageList(convId?: string): Promise<MessageRow[]> {
  const q = db.select({
    id: message.id, conv_id: message.convId, sender_id: message.senderId,
    sender_name: message.senderName, content: message.content, content_type: message.contentType,
    is_self: message.isSelf, msg_seq: message.msgSeq, client_msg_id: message.clientMsgId, ts: message.ts,
  }).from(message).$dynamic();
  if (convId) return q.where(eq(message.convId, convId)).orderBy(message.id) as unknown as Promise<MessageRow[]>;
  return q.orderBy(message.id) as unknown as Promise<MessageRow[]>;
}

// ============ 审计 ============

export async function auditLog(actor: string, action: string, detail: Record<string, unknown> | null = null): Promise<void> {
  await db.insert(audit).values({ actor, action, detail: detail ? JSON.stringify(detail) : null });
}

/** 生产 audit 已对齐代码 schema（ts + TEXT detail，drizzle 0000 迁移），探测分支退役。 */
export async function auditRecent(limit = 30): Promise<Array<{ actor: string; action: string; detail: unknown; ts: string }>> {
  const rows = await db.select({
    actor: audit.actor, action: audit.action, detail: audit.detail, ts: audit.ts,
  }).from(audit).orderBy(desc(audit.id)).limit(limit) as unknown as Array<{ actor: string; action: string; detail: unknown; ts: string }>;
  return rows.map((r) => {
    let detail: unknown = r.detail;
    if (typeof detail === "string") {
      try { detail = JSON.parse(detail); } catch { /* 保持原字符串 */ }
    }
    return { actor: r.actor, action: r.action, detail, ts: String(r.ts) };
  });
}

// ============ 术语 ============

export interface TermRow { id: number; term: string; meaning: string; source: string; created_at: Date | string; }

export async function listTermDicts(): Promise<TermRow[]> {
  const rows = await db.select({
    id: term.id, term: term.term, meaning: term.meaning, source: term.source, created_at: term.createdAt,
  }).from(term).orderBy(desc(term.id)) as unknown as Promise<TermRow[]>;
  return rows;
}

// 保留 sql 模板逃生舱引用（复杂聚合在 stats.ts 使用）
export const _sql = sql;

File: im-ai-office/imai/repos.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""数据访问层：集中高频/复用查询（SQL 语义与 core.py 1:1）

约定：
- 只放被多模块复用或结构复杂的查询；简单单行写允许留在服务层原样直写
- 全部函数接收 con（连接由调用方管理），不自行开关连接
"""
from imai.db import _rows, take_id


# ============ 人 / 别名 ============

def find_persons_by_alias(con, name):
    """别名 → 候选人列表 [dict(id, real_name, flower_name)]（双方言统一 dict 行）"""
    c = con.cursor()
    c.execute("SELECT p.id, p.real_name, p.flower_name FROM alias a JOIN person p ON p.id=a.person_id WHERE a.name=?", (name,))
    return _rows(c)


def distinct_alias_names(con):
    c = con.cursor()
    c.execute("SELECT DISTINCT name FROM alias")
    return [r["name"] for r in _rows(c)]


def alias_label_rows(con):
    """DISTINCT 别名-正名-花名（注入与溯源共用），dict 行。"""
    c = con.cursor()
    c.execute("SELECT DISTINCT a.name, p.real_name, p.flower_name FROM alias a JOIN person p ON p.id=a.person_id")
    return _rows(c)


def insert_alias_if_absent(con, person_id, name) -> bool:
    """幂等插入别名；返回是否实际新增。"""
    c = con.cursor()
    c.execute("SELECT 1 FROM alias WHERE person_id=? AND name=?", (person_id, name))
    if not c.fetchone():
        c.execute("INSERT INTO alias(person_id, name) VALUES(?,?)", (person_id, name))
        con.commit()
        return True
    return False


# ============ 任务 ============

def insert_task(con, content, creator, assignee, deadline, status, confidence,
                source_msg, pending_meta=None):
    """插入任务，返回自增 id。pending_meta 由调用方保证 JSON 字符串。"""
    if pending_meta is not None:
        c = con.cursor()
        c.execute(
            "INSERT INTO task(content,creator,assignee,deadline,status,confidence,source_msg,pending_meta)"
            " VALUES(?,?,?,?,?,?,?,?) RETURNING id",
            (content, creator, assignee, deadline, status, confidence, source_msg, pending_meta))
    else:
        c = con.cursor()
        c.execute(
            "INSERT INTO task(content,creator,assignee,deadline,status,confidence,source_msg)"
            " VALUES(?,?,?,?,?,?,?) RETURNING id",
            (content, creator, assignee, deadline, status, confidence, source_msg))
    tid = take_id(c)
    con.commit()
    return tid


def get_task_dict(con, task_id):
    c = con.cursor()
    c.execute("SELECT * FROM task WHERE id=?", (task_id,))
    row = c.fetchone()
    return dict(row) if row else None


def list_task_dicts(con, status=None):
    """默认排除 cancelled（迭代2 B1 终态，看板不展示）；status 参数优先。"""
    c = con.cursor()
    if status:
        c.execute("SELECT * FROM task WHERE status=? ORDER BY id DESC", (status,))
    else:
        c.execute("SELECT * FROM task WHERE status != 'cancelled' ORDER BY id DESC")
    return _rows(c)


def latest_pending_assignee_for_creator(con, creator):
    """取该发送者最近一条 pending_assignee 任务。"""
    c = con.cursor()
    c.execute("SELECT * FROM task WHERE creator=? AND status='pending_assignee' ORDER BY id DESC LIMIT 1", (creator,))
    row = c.fetchone()
    return dict(row) if row else None


def latest_pending_assignee_by_dm_taskid(con, sender_id):
    """从 ai_dm 最近带 task_id 的记录回查 pending_assignee 任务。"""
    task = None
    c = con.cursor()
    c.execute("SELECT * FROM ai_dm WHERE sender_id=? AND task_id IS NOT NULL ORDER BY id DESC LIMIT 1", (sender_id,))
    row = c.fetchone()
    if row:
        tid = row["task_id"]
        c.execute("SELECT * FROM task WHERE id=? AND status='pending_assignee' ORDER BY id DESC LIMIT 1", (tid,))
        t2 = c.fetchone()
        task = dict(t2) if t2 else None
    return task


# ============ 消息表（本地 message）============

def message_add(con, conv_id, sender_id, sender_name, content, is_self=0,
                msg_seq=None, client_msg_id=None, content_type=101):
    """记录一条消息（自己发或收到的）。返回自增 id。
    幂等：同 conv 内相同 client_msg_id 已存在时直接返回既有 id，
    防 SDK 重连重投递导致重复入库/重复 AI（2026-08-30 实证）。"""
    c = con.cursor()
    if client_msg_id:
        c.execute("SELECT id FROM message WHERE conv_id=? AND client_msg_id=? LIMIT 1",
                  (conv_id, client_msg_id))
        row = c.fetchone()
        if row:
            con.commit()
            return row[0]
    c.execute(
        "INSERT INTO message(conv_id, sender_id, sender_name, content, is_self, msg_seq, client_msg_id, content_type) VALUES(?,?,?,?,?,?,?,?) RETURNING id",
        (conv_id, sender_id, sender_name, content, is_self, msg_seq, client_msg_id, content_type))
    mid = take_id(c)
    con.commit()
    return mid


def message_list(con, conv_id=None):
    """取某会话的消息历史，按 id 升序。"""
    c = con.cursor()
    if conv_id:
        c.execute("SELECT * FROM message WHERE conv_id=? ORDER BY id ASC", (conv_id,))
    else:
        c.execute("SELECT * FROM message ORDER BY id ASC")
    return _rows(c)


# ============ 审计 ============

def audit_log(con, actor, action, detail=None):
    """审计：所有关键 AI/人工动作留痕。actor: ai | user:<id> | system。detail 序列化为 JSON 字符串。"""
    import json
    c = con.cursor()
    c.execute("INSERT INTO audit(actor,action,detail) VALUES(?,?,?)",
              (actor, action, json.dumps(detail, ensure_ascii=False) if detail is not None else None))
    con.commit()


def audit_recent(con, limit=30):
    c = con.cursor()
    # PG 无 rowid：用自增 id；SQLite 保留 rowid。
    # 时间列运行时探测（交接文档 #11）：代码 schema=ts，生产旧库=created_at（先例 stats._audit_time_col）
    from imai.db import BACKEND
    if BACKEND == "postgres":
        c.execute("SELECT column_name AS col FROM information_schema.columns "
                  "WHERE table_name='audit' AND column_name IN ('ts','created_at')")
        names = {r["col"] for r in c.fetchall()}
        tcol = "ts" if "ts" in names else ("created_at" if "created_at" in names else None)
        if not tcol:
            raise RuntimeError("audit 表缺少时间列（ts/created_at）")
        c.execute(f"SELECT actor,action,detail,{tcol} AS ts FROM audit ORDER BY id DESC LIMIT %s", (limit,))
    else:
        c.execute("SELECT actor,action,detail,ts FROM audit ORDER BY rowid DESC LIMIT ?", (limit,))
    return _rows(c)


# ============ 术语 / 团队记忆 ============

def list_term_dicts(con):
    c = con.cursor()
    c.execute("SELECT * FROM term ORDER BY id DESC")
    return _rows(c)
